pkg/operation/botanist/clusteridentity/clusteridentity.go

```diff
@@ -0,0 +1,124 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package clusteridentity
+
+import (
+	"context"
+	"fmt"
+
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	v1beta1constants "github.com/gardener/gardener/pkg/apis/core/v1beta1/constants"
+	"github.com/gardener/gardener/pkg/operation/botanist/component"
+	kutil "github.com/gardener/gardener/pkg/utils/kubernetes"
+
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+type shootClusterIdentity struct {
+	clusterIdentity *string
+	seedNamespace   string
+	namespace       string
+	name            string
+	uid             string
+
+	seedClient   client.Client
+	gardenClient client.Client
+
+	logger *logrus.Entry
+}
+
+func New(
+	clusterIdentity *string,
+	name, namespace, seedNamespace, uid string,
+	seedClient, gardenClient client.Client,
+	logger *logrus.Entry,
+) component.Deployer {
+	return &shootClusterIdentity{
+		clusterIdentity: clusterIdentity,
+		name:            name,
+		namespace:       namespace,
+		uid:             uid,
+		seedNamespace:   seedNamespace,
+		seedClient:      seedClient,
+		gardenClient:    gardenClient,
+		logger:          logger,
+	}
+}
+
+func (c *shootClusterIdentity) Deploy(ctx context.Context) error {
+	if c.clusterIdentity == nil {
+		clusterIdentity, err := c.determineShootClusterIdentity(ctx)
+		if err != nil {
+			return err
+		}
+
+		shoot := &gardencorev1beta1.Shoot{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      c.name,
+				Namespace: c.namespace,
+			},
+		}
+
+		c.clusterIdentity = &clusterIdentity
+		patch := []byte(fmt.Sprintf(`{"status":{"clusterIdentity": "%s"}`, clusterIdentity))
+		if err := c.gardenClient.Status().Patch(ctx, shoot, client.RawPatch(types.StrategicMergePatchType, patch)); err != nil {
+			return err
+		}
+	}
+
+	shootClusterIdentity := &corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      v1beta1constants.ClusterIdentity,
+			Namespace: c.seedNamespace,
+		},
+	}
+
+	_, err := controllerutil.CreateOrUpdate(ctx, c.seedClient, shootClusterIdentity, func() error {
+		shootClusterIdentity.Data = map[string]string{v1beta1constants.ClusterIdentity: *c.clusterIdentity}
+		return nil
+	})
+	return err
+}
+
+// Destroy returns nil
+func (c *shootClusterIdentity) Destroy(ctx context.Context) error {
+	return nil
+}
+
+// determineShootClusterIdentity determines the identity of a Shoot cluster, in cases that the identity exists it returns its value, else computes the identity and returns it.
+func (c *shootClusterIdentity) determineShootClusterIdentity(ctx context.Context) (string, error) {
```

Comment:
Shouldn't this return the `.status.clusterIdentity` of the `Shoot` if set instead of recomputing it?

Reply:
Yes, this method is called if the status.clusterIdentity is missing. This is checked in  `Deploy` where c.clusterIdentity is the .status.clusterIdentity

Reply:
I see, thank you.